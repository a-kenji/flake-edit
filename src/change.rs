#[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize)]
pub enum Change {
    #[default]
    None,
    Add {
        id: Option<String>,
        uri: Option<String>,
    },
    Remove {
        id: String,
    },
    Pin {
        id: String,
    },
    Change {
        id: Option<String>,
        ref_or_rev: Option<String>,
    },
}

impl Change {
    pub fn id(&self) -> Option<String> {
        match self {
            Change::None => None,
            Change::Add { id, .. } => id.clone(),
            Change::Remove { id } => Some(id.clone()),
            Change::Change { id, .. } => id.clone(),
            Change::Pin { id } => Some(id.clone()),
        }
    }
    pub fn is_remove(&self) -> bool {
        matches!(self, Change::Remove { .. })
    }
    pub fn is_some(&self) -> bool {
        !matches!(self, Change::None)
    }
}
