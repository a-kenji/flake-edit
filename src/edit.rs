use std::collections::HashMap;

use crate::change::Change;
use crate::error::FlakeEditError;
use crate::input::Input;
use crate::walk::Walker;

pub struct FlakeEdit {
    changes: Vec<Change>,
    walker: Walker,
}

#[derive(Default, Debug)]
pub enum Outputs {
    #[default]
    None,
    // needs strict inputs to output mapping
    Multiple(Vec<String>),
    // contains ...
    Any(Vec<String>),
}

#[derive(Default, Debug)]
pub enum OutputChange {
    #[default]
    None,
    Add(String),
    Remove(String),
}

impl FlakeEdit {
    pub fn new(changes: Vec<Change>, walker: Walker) -> Self {
        Self { changes, walker }
    }

    pub fn from(stream: &str) -> Result<Self, FlakeEditError> {
        let walker = Walker::new(stream);
        Ok(Self::new(Vec::new(), walker))
    }

    pub fn changes(&self) -> &[Change] {
        self.changes.as_ref()
    }

    pub fn add_change(&mut self, change: Change) {
        self.changes.push(change);
    }

    pub fn curr_list(&self) -> &HashMap<String, Input> {
        &self.walker.inputs
    }

    /// Will walk and then list the inputs, for listing the current inputs,
    /// use `curr_list()`.
    pub fn list(&mut self) -> &HashMap<String, Input> {
        self.walker.inputs.clear();
        assert!(self.walker.walk(&Change::None).is_none());
        &self.walker.inputs
    }
    /// Apply a specific change to a walker, on some inputs it will need to walk
    /// multiple times, will error, if the edit could not be applied successfully.
    pub fn apply_change(&mut self, change: Change) -> Result<Option<String>, FlakeEditError> {
        match change {
            Change::None => Ok(None),
            //TODO: Add outputs, if needed.
            Change::Add { .. } => {
                if let Some(maybe_changed_node) = self.walker.walk(&change.clone()) {
                    let outputs = self.walker.list_outputs();
                    match outputs {
                        Outputs::Multiple(out) => {
                            let id = change.id().unwrap().to_string();
                            if !out.contains(&id) {
                                self.walker.root = maybe_changed_node.clone();
                                if let Some(maybe_changed_node) =
                                    self.walker.change_outputs(OutputChange::Add(id))
                                {
                                    return Ok(Some(maybe_changed_node.to_string()));
                                }
                            }
                        }
                        Outputs::None | Outputs::Any(_) => {}
                    }
                    Ok(Some(maybe_changed_node.to_string()))
                } else {
                    self.walker.add_toplevel = true;
                    let maybe_changed_node = self.walker.walk(&change);
                    Ok(maybe_changed_node.map(|n| n.to_string()))
                }
            }
            Change::Remove { .. } => {
                // If we remove a node, it could be a flat structure,
                // we want to remove all of the references to its toplevel.
                let mut res = None;
                while let Some(changed_node) = self.walker.walk(&change) {
                    if res == Some(changed_node.clone()) {
                        // TODO: Sanity check, turn into proper error.
                        break;
                    }
                    res = Some(changed_node.clone());
                    self.walker.root = changed_node.clone();
                }
                // Removed nodes should be removed from the outputs
                let outputs = self.walker.list_outputs();
                match outputs {
                    Outputs::Multiple(out) => {
                        let id = change.id().unwrap().to_string();
                        if out.contains(&id) {
                            if let Some(changed_node) =
                                self.walker.change_outputs(OutputChange::Remove(id))
                            {
                                res = Some(changed_node.clone());
                                self.walker.root = changed_node.clone();
                            }
                        }
                    }
                    Outputs::None | Outputs::Any(_) => {}
                }
                Ok(res.map(|n| n.to_string()))
            }
            Change::Pin { .. } => todo!(),
            Change::Change { .. } => {
                if let Some(maybe_changed_node) = self.walker.walk(&change) {
                    Ok(Some(maybe_changed_node.to_string()))
                } else {
                    panic!("No change");
                }
            }
        }
    }

    pub fn walker(&self) -> &Walker {
        &self.walker
    }
}
