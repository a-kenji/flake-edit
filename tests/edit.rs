use flake_edit::change::Change;
use flake_edit::edit::FlakeEdit;
use flake_edit::walk::Walker;

fn load_fixtures(name: &str) -> (String, String) {
    let dir = env!("CARGO_MANIFEST_DIR");
    let flake_nix =
        std::fs::read_to_string(format!("{dir}/tests/fixtures/{name}.flake.nix")).unwrap();
    let flake_lock =
        std::fs::read_to_string(format!("{dir}/tests/fixtures/{name}.flake.lock")).unwrap();
    (flake_nix, flake_lock)
}

#[derive(serde::Serialize)]
struct Info {
    flake_nix: String,
    changes: Vec<Change>,
}

impl Info {
    fn new(flake_nix: String, changes: Vec<Change>) -> Self {
        Self { flake_nix, changes }
    }
}

#[test]
fn root_load() {
    let (_flake, _lock) = load_fixtures("root");
}

#[test]
fn root_edit_list() {
    let (flake, _lock) = load_fixtures("root");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let info = Info::new("".into(), vec![]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_yaml_snapshot!(flake_edit.list());
    });
}
#[test]
fn root_add_toplevel_id_uri() {
    let (flake, _lock) = load_fixtures("root");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let change = Change::Add {
        id: Some("vmsh".to_owned()),
        uri: Some("github:mic92/vmsh".to_owned()),
    };
    let info = Info::new("".into(), vec![change.clone()]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_snapshot!(flake_edit.apply_change(change).unwrap().unwrap());
    });
}
#[test]
fn root_remove_toplevel_uri() {
    let (flake, _lock) = load_fixtures("root");
    let mut walker = Walker::new(&flake);
    let change = Change::Remove {
        id: "nixpkgs".to_owned(),
    };
    let info = Info::new("".into(), vec![change.clone()]);
    let change = walker.walk(&change).unwrap();
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_snapshot!(change.to_string());
    });
}
// #[test]
// fn root_remove_toplevel_input_multiple() {
//     let (flake, _lock) = load_fixtures("root");
//     let mut walker = Walker::new(&flake);
//     let change = Change::Remove {
//         id: "crane".to_owned(),
//     };
//     let info = Info::new("".into(), vec![change.clone()]);
//     let change = walker.walk(&change).unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
#[test]
fn root_alt_list() {
    let (flake, _lock) = load_fixtures("root_alt");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let info = Info::new("".into(), vec![]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_yaml_snapshot!(flake_edit.list());
    });
}
// #[test]
// fn root_alt_add_toplevel_id_uri() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake);
//     let change = Change::Add {
//         id: Some("vmsh".to_owned()),
//         uri: Some("github:mic92/vmsh".to_owned()),
//     };
//     let info = Info::new("".into(), vec![change.clone()]);
//     let change = walker.walk(&change).unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
// #[test]
// fn root_alt_remove_toplevel_uri() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake);
//     let change = Change::Remove {
//         id: "nixpkgs".to_owned(),
//     };
//     let info = Info::new("".into(), vec![change.clone()]);
//     let change = walker.walk(&change).unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
// #[test]
// fn root_alt_remove_toplevel_input_multiple() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake);
//     let change = Change::Remove {
//         id: "crane".to_owned(),
//     };
//     let info = Info::new("".into(), vec![change.clone()]);
//     let change = walker.walk(&change).unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
//
#[test]
fn root_toplevel_nesting_list() {
    let (flake, _lock) = load_fixtures("toplevel_nesting");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let info = Info::new("".into(), vec![]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_yaml_snapshot!(flake_edit.list());
    });
}
// #[test]
// fn root_alt_add_toplevel_id_uri() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake).unwrap();
//     let change = Change::Add {
//         id: Some("vmsh".to_owned()),
//         uri: Some("github:mic92/vmsh".to_owned()),
//     };
//     walker.changes.push(change.clone());
//     let info = Info::new("".into(), vec![change]);
//     let change = walker.walk().unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
// #[test]
// fn root_alt_remove_toplevel_uri() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake).unwrap();
//     let change = Change::Remove {
//         id: "nixpkgs".to_owned(),
//     };
//     walker.changes.push(change.clone());
//     let info = Info::new("".into(), vec![change]);
//     let change = walker.walk().unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
// #[test]
// fn root_alt_remove_toplevel_input_multiple() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake).unwrap();
//     let change = Change::Remove {
//         id: "crane".to_owned(),
//     };
//     walker.changes.push(change.clone());
//     let info = Info::new("".into(), vec![change]);
//     let change = walker.walk().unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }

#[test]
fn completely_flat_toplevel_list() {
    let (flake, _lock) = load_fixtures("completely_flat_toplevel");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let info = Info::new("".into(), vec![]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_yaml_snapshot!(flake_edit.list());
    });
}

#[test]
fn completely_flat_toplevel_alt_list() {
    let (flake, _lock) = load_fixtures("completely_flat_toplevel_alt");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let info = Info::new("".into(), vec![]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_yaml_snapshot!(flake_edit.list());
    });
}
#[test]
fn completely_flat_toplevel_add_id_uri() {
    let (flake, _lock) = load_fixtures("completely_flat_toplevel");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let change = Change::Add {
        id: Some("vmsh".to_owned()),
        uri: Some("mic92/vmsh".to_owned()),
    };
    let info = Info::new("".into(), vec![change.clone()]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_snapshot!(flake_edit.apply_change(change).unwrap().unwrap());
    });
}
#[test]
fn completely_flat_toplevel_rm_toplevel() {
    let (flake, _lock) = load_fixtures("completely_flat_toplevel");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let change = Change::Remove {
        id: "nixpkgs".to_owned(),
    };
    let info = Info::new("".into(), vec![change.clone()]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_snapshot!(flake_edit.apply_change(change).unwrap().unwrap());
    });
}
#[test]
fn completely_flat_toplevel_rm_toplevel_muliple() {
    let (flake, _lock) = load_fixtures("completely_flat_toplevel");
    let mut flake_edit = FlakeEdit::from(&flake).unwrap();
    let change = Change::Remove {
        id: "crane".to_owned(),
    };
    let info = Info::new("".into(), vec![change.clone()]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_snapshot!(flake_edit.apply_change(change).unwrap().unwrap());
    });
}
// #[test]
// fn root_alt_remove_toplevel_uri() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake).unwrap();
//     let change = Change::Remove {
//         id: "nixpkgs".to_owned(),
//     };
//     walker.changes.push(change.clone());
//     let info = Info::new("".into(), vec![change]);
//     let change = walker.walk().unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
// #[test]
// fn root_alt_remove_toplevel_input_multiple() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake).unwrap();
//     let change = Change::Remove {
//         id: "crane".to_owned(),
//     };
//     walker.changes.push(change.clone());
//     let info = Info::new("".into(), vec![change]);
//     let change = walker.walk().unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }

#[test]
fn one_level_nesting_flat() {
    let (flake, _lock) = load_fixtures("one_level_nesting_flat");
    let mut walker = Walker::new(&flake);
    walker.walk(&Change::None);
    let info = Info::new("".into(), vec![]);
    insta::with_settings!({sort_maps => true, info => &info}, {
        insta::assert_yaml_snapshot!(walker.inputs);
    });
}
// #[test]
// fn root_alt_add_toplevel_id_uri() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake).unwrap();
//     let change = Change::Add {
//         id: Some("vmsh".to_owned()),
//         uri: Some("github:mic92/vmsh".to_owned()),
//     };
//     walker.changes.push(change.clone());
//     let info = Info::new("".into(), vec![change]);
//     let change = walker.walk().unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
// #[test]
// fn root_alt_remove_toplevel_uri() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake).unwrap();
//     let change = Change::Remove {
//         id: "nixpkgs".to_owned(),
//     };
//     walker.changes.push(change.clone());
//     let info = Info::new("".into(), vec![change]);
//     let change = walker.walk().unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
// #[test]
// fn root_alt_remove_toplevel_input_multiple() {
//     let (flake, _lock) = load_fixtures("root_alt");
//     let mut walker = Walker::new(&flake).unwrap();
//     let change = Change::Remove {
//         id: "crane".to_owned(),
//     };
//     walker.changes.push(change.clone());
//     let info = Info::new("".into(), vec![change]);
//     let change = walker.walk().unwrap();
//     insta::with_settings!({sort_maps => true, info => &info}, {
//         insta::assert_snapshot!(change.to_string());
//     });
// }
