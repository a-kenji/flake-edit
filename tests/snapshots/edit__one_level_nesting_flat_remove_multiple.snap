---
source: tests/edit.rs
expression: flake_edit.apply_change(change).unwrap().unwrap()
info:
  flake_nix: ""
  changes: []
---
{
  description = "Edit your flake inputs with ease";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    flake-utelinos.url = "github:numtide/flake-utils";
    crane.url = "github:ipetkov/crane";
    crane.inputs.nixpkgs.follows = "nixpkgs";
    crane.inputs.rust-overlay.follows = "rust-overlay";
    crane.inputs.flake-utils.follows = "flake-utils";
  };

  outputs = {
    self,
    nixpkgs,
    flake-utils,
    flake-utelinos,
    rust-overlay,
    crane,
  }: {};
}
